src/main/java/org/opensearch/knn/index/codec/KNN990Codec/KNN990QuantizationStateWriter.java

```diff
@@ -0,0 +1,115 @@
+import org.apache.lucene.index.SegmentWriteState;
+import org.apache.lucene.store.IndexOutput;
+import org.opensearch.knn.common.KNNConstants;
+import org.opensearch.knn.quantization.models.quantizationState.QuantizationState;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Writes quantization states to off heap memory
+ */
+public final class KNN990QuantizationStateWriter {
+
+    private final IndexOutput output;
+    private List<FieldQuantizationState> fieldQuantizationStates = new ArrayList<>();
+
+    /**
+     * Constructor
+     * Overall file format for writer:
+     * Header
+     * QS1 state bytes
+     * QS2 state bytes
+     * Number of quantization states
+     * QS1 field number
+     * QS1 state bytes length
+     * QS1 position of state bytes
+     * QS2 field number
+     * QS2 state bytes length
+     * QS2 position of state bytes
+     * Position of index section (where QS1 field name is located)
+     * -1 (marker)
+     * Footer
+     * @param segmentWriteState segment write state containing segment information
+     * @throws IOException exception could be thrown while creating the output
+     */
+    public KNN990QuantizationStateWriter(SegmentWriteState segmentWriteState) throws IOException {
+        String quantizationStateFileName = IndexFileNames.segmentFileName(
+            segmentWriteState.segmentInfo.name,
+            segmentWriteState.segmentSuffix,
+            KNNConstants.QUANTIZATION_STATE_FILE_SUFFIX
+        );
+
+        output = segmentWriteState.directory.createOutput(quantizationStateFileName, segmentWriteState.context);
+    }
+
+    /**
+     * Writes an index header
+     * @param segmentWriteState state containing segment information
+     * @throws IOException exception could be thrown while writing header
+     */
+    public void writeHeader(SegmentWriteState segmentWriteState) throws IOException {
+        CodecUtil.writeIndexHeader(
+            output,
+            "NativeEngines99KnnVectorsFormatQSData",
+            0,
+            segmentWriteState.segmentInfo.getId(),
+            segmentWriteState.segmentSuffix
+        );
+    }
+
+    /**
+     * Writes a quantization state as bytes
+     *
+     * @param fieldNumber field number
+     * @param quantizationState quantization state
+     * @throws IOException could be thrown while writing
+     */
+    public void writeState(int fieldNumber, QuantizationState quantizationState) throws IOException {
+        byte[] stateBytes = quantizationState.toByteArray();
+        long position = output.getFilePointer();
+        output.writeBytes(stateBytes, stateBytes.length);
+        fieldQuantizationStates.add(new FieldQuantizationState(fieldNumber, stateBytes, position));
+    }
+
+    /**
+     * Writes index footer and other index information for parsing later
+     * @throws IOException could be thrown while writing
+     */
+    public void writeFooter() throws IOException {
```

Comment:
thinking out loud here,

There is a small risk here that writing of footer is dependent on client. To safe guard that the footer is written you can check if its written in close (if this class implement closeable). One of the downsides I see there is that if its used with try with resources the footer will be written even in exception cases and I am not sure if thats considered as a corrupt file